### ISSUE TITLE
MinGW32 tests hang sometimes

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1573]
MinGW32 tests hang sometimes and the CI job times out. It seems to be completely unpredictable which test will hang.

Here are some examples:

 - https://github.com/igraph/igraph/runs/1571724909
 - https://github.com/igraph/igraph/runs/1571967929

**Comment by szhorvat** [comment]
@jannick0 @jankatins Have either of you seen anything like this when working with MinGW?

**Comment by szhorvat** [comment]
The MinGW tests are now switched to use internal BLAS/LAPACK to see if the hangs have anything to do with OpenBLAS.

**Comment by jannick0** [comment]
> @jannick0 @jankatins Have either of you seen anything like this when working with MinGW?

Actually no. If switching libraries does not help, you could check
- if this happens locally, too
- if it always hangs at the last test (by permuting the test order)
- use MSYS's cmake which - as of now - is at version 3.18.4 (while MINGW32's cmake is at 3.19.1)

(I have not yet had a look at the entire cmake machinery in the repo.)

**Comment by jannick0** [comment]
> Here are some examples:

Thanks. I am not so deep in cmake, but what is the interpretation that the hanging test changed from one run to the other - or did a commit kick in here? ... just wondering to help.  
  
>     * https://github.com/igraph/igraph/runs/1571724909
~~~
274/276 Test #275: igraph_is_graphical .......................   Passed    0.07 sec
         Start 276: igraph_is_bigraphical
275/276 Test #276: igraph_is_bigraphical .....................   Passed    0.07 sec
276/276 Test #111: igraph_transitivity .......................***Timeout 1500.00 sec
~~~ 
>     * https://github.com/igraph/igraph/runs/1571967929
~~~
274/276 Test #275: igraph_is_graphical .......................   Passed    0.09 sec
        Start 276: igraph_is_bigraphical
275/276 Test #276: igraph_is_bigraphical .....................   Passed    0.08 sec
276/276 Test  #55: igraph_barabasi_game ......................***Timeout 1500.00 sec
~~~

**Comment by jannick0** [comment]
> The MinGW tests are now switched to use internal BLAS/LAPACK to see if the hangs have anything to do with OpenBLAS.

[That run](https://github.com/igraph/igraph/runs/1572138018) succeeded.

**Comment by vtraag** [comment]
I experienced a similar problem with OpenBLAS when building on MSVC, see https://github.com/igraph/igraph/issues/1491. Note that a more  recent version of OpenBLAS (0.3.10) I did no longer encounter such issues. Perhaps the version in MinGW32 is lagging behind?

**Comment by jannick0** [comment]
This is unlikely: The versions of MINGW32 and MINGW64 packages are always in sync. OpenBlas is at 0.3.12.

**Comment by vtraag** [comment]
Then I'm not sure what is causing the problem. Given that it now ran successfully, it suggests there might still be some problem with OpenBLAS on MinGW, even with more recent versions.

**Comment by szhorvat** [comment]
> [That run](https://github.com/igraph/igraph/runs/1572138018) succeeded.

The problem is intermittent though, so this doesn't tell us much _yet_. Let us see if it comes back. I just triggered a re-run and will do it a few more times.

> I am not so deep in cmake, but what is the interpretation that the hanging test changed from one run to the other - or did a commit kick in here? ... just wondering to help.

That's the strange thing ... No, there was no commit. The problem is random. However, I have seen `igraph_barabasi_game` get stuck more than once, which is suspicious.

I have seen a hang when I accidentally removed the random seed from a test. It ended up generating a random graph which was not valid input for the tested functionality. I don't believe that we have _any_ non-deterministic tests (i.e. the seed is always set), but I will double-check anyway.

> I experienced a similar problem with OpenBLAS when building on MSVC

This is why I got rid of OpenBLAS temporarily. However, the tests that hang do not use OpenBLAS.

My uninformed theory: even if it doesn't use OpenBLAS, it still links to it (to be verified!). Maybe OpenBLAS has some faulty startup code. As far as I'm aware, OpenBLAS is the only multi-threaded library we link to. There are complaints about race conditions in OpenBLAS online.

**Comment by vtraag** [comment]
The problem with MSVC was also completely random, it sometimes got stuck on `igraph_version` :). The problem seems to be in the exit procedure (see https://github.com/igraph/igraph/issues/1490#issuecomment-715383082), when it probably does a join/wait for threads, which apparently goes wrong.

**Comment by ntamas** [comment]
Tests without OpenBLAS passed four times in a row without getting stuck, so I consider this fixed.

**Comment by szhorvat** [comment]
There's something more going on here. Today I re-enabled building with external BLAS/LAPACK but I used NetLib LAPACK. It seems to be hanging again.

Also, I noticed that `igraph_barabasi_game` does not actually set a random seed. The randomness may be coming from there. Still, I don't see why it would hang only on 32-bit, especially considering that on 64-bit it does not hang, and that Windows, both `int` and `long` are 32-bit _both_ in 32- and 64-bit mode.
